gtsam/slam/SmartProjectionRigFactor.h

```diff
@@ -0,0 +1,366 @@
+   * @param params parameters for the smart projection factors
+   */
+  SmartProjectionRigFactor(const SharedNoiseModel& sharedNoiseModel,
+                           const Cameras& cameraRig,
+                           const SmartProjectionParams& params =
+                               SmartProjectionParams())
+      : Base(sharedNoiseModel, params),
+        cameraRig_(cameraRig) {
+    // use only configuration that works with this factor
+    Base::params_.degeneracyMode = gtsam::ZERO_ON_DEGENERACY;
+    Base::params_.linearizationMode = gtsam::HESSIAN;
+  }
+
+  /**
+   * Constructor
+   * @param sharedNoiseModel isotropic noise model for the 2D feature measurements
+   * @param camera single camera (fixed poses wrt body and intrinsics)
+   * @param params parameters for the smart projection factors
+   */
+  SmartProjectionRigFactor(const SharedNoiseModel& sharedNoiseModel,
+                           const Camera& camera,
+                           const SmartProjectionParams& params =
+                               SmartProjectionParams())
+      : Base(sharedNoiseModel, params) {
+    // use only configuration that works with this factor
+    Base::params_.degeneracyMode = gtsam::ZERO_ON_DEGENERACY;
+    Base::params_.linearizationMode = gtsam::HESSIAN;
+    cameraRig_.push_back(camera);
+  }
+
+  /** Virtual destructor */
+  ~SmartProjectionRigFactor() override {
+  }
+
+  /**
+   * add a new measurement, corresponding to an observation from pose "poseKey"
+   * and taken from the camera in the rig identified by "cameraId"
+   * @param measured 2-dimensional location of the projection of a
+   * single landmark in a single view (the measurement)
+   * @param poseKey key corresponding to the body pose of the camera taking the measurement
+   * @param cameraId ID of the camera in the rig taking the measurement (default 0)
+   */
+  void add(const Point2& measured, const Key& poseKey, const size_t cameraId = 0) {
+    // store measurement and key
+    this->measured_.push_back(measured);
+    this->nonUniqueKeys_.push_back(poseKey);
+
+    //  also store keys in the keys_ vector: these keys are assumed to be unique, so we avoid duplicates here
+    if (std::find(this->keys_.begin(), this->keys_.end(), poseKey) == this->keys_.end())
+      this->keys_.push_back(poseKey);  // add only unique keys
+
+    // store id of the camera taking the measurement
+    cameraIds_.push_back(cameraId);
+  }
+
+  /**
+   * Variant of the previous "add" function in which we include multiple measurements
+   * @param measurements vector of the 2m dimensional location of the projection
+   * of a single landmark in the m views (the measurements)
+   * @param poseKeys keys corresponding to the body poses of the cameras taking the measurements
+   * @param cameraIds IDs of the cameras in the rig taking each measurement (same order as the measurements)
+   */
+  void add(const Point2Vector& measurements, const KeyVector& poseKeys,
```

Comment:
`const MEASUREMENTS& measurements` here as well

Reply:
actually I accepted this comment